feat: add `array/base/count-ifs`

headlessNode opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

## Description

> What is the purpose of this pull request?

This pull request:

-   adds `array/base/count-ifs` including benchmarks, tests and documentation.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/18

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add array/base/count-ifs

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/array/base/count-ifs/README.md:
````diff
@@ -0,0 +1,149 @@
+// returns 1
+```
+
+The `predicate` function is provided three arguments:
````
```suggestion
Each predicate function is provided three arguments:
```

kgryte commented on lib/node_modules/@stdlib/array/base/count-ifs/README.md:
```diff
@@ -0,0 +1,149 @@
+});
+// returns <Int32Array>
+
+var out = countIfs( x, naryFunction( isPositiveInteger, 1 ), y, naryFunction( isNegativeInteger, 1 ) ); // eslint-disable-line max-len
```
You need to disable lint rules via an HTML comment placed before this code block. See other READMEs.

kgryte commented on lib/node_modules/@stdlib/array/base/count-ifs/benchmark/benchmark.length.js:
```diff
@@ -0,0 +1,123 @@
+* @returns {Function} benchmark function
+*/
+function createBenchmark( len ) {
+	var x0 = bernoulli( len, 0.5, {
```
Based on your predicates, I think you want `uniform( len, -1.0, 1.0, {...} )` here. Bernoulli can only be `0` or `1`, but `predicate1` is checking for less than `0`.

kgryte commented on lib/node_modules/@stdlib/array/base/count-ifs/docs/repl.txt:
```diff
@@ -0,0 +1,39 @@
+    Counts the number of elements in the input arrays which pass the tests
+    implemented by the predicate functions.
+
+    The predicate function is provided three arguments:
```
```suggestion
    Each predicate function is provided three arguments:
```

kgryte commented on lib/node_modules/@stdlib/array/base/count-ifs/docs/types/index.d.ts:
```diff
@@ -0,0 +1,149 @@
+* var n = countIfs( x, predicate );
+* // returns 3
+*/
+declare function countIfs<U = unknown>( x: Collection<U> | AccessorArrayLike<U>, predicate: Predicate<U> ): number;
```
Create a type `InputArray<U> = Collection<U> | AccessorArrayLike<U>`. Then, see about doing

```
declare function countIfs<T = unknown, U extends InputArray<T>>( x: U, predicate: Predicate<T, U> ): number;
```

You'll then need to update the predicates accordingly. In short, we want to preserve type information for the `arr` argument. Otherwise, right now, we lose type specificity in the callback.

kgryte replied:
Use consistent naming. s/x/x0/ and s/predicate/predicate0/.

kgryte commented on lib/node_modules/@stdlib/array/base/count-ifs/docs/types/index.d.ts:
```diff
@@ -0,0 +1,149 @@
+* var n = countIfs( x0, predicate0, x1, predicate1 );
+* // returns 2
+*/
+declare function countIfs<U = unknown>( x0: Collection<U> | AccessorArrayLike<U>, predicate0: Predicate<U>, x1: Collection<U> | AccessorArrayLike<U>, predicate1: Predicate<U>, x2: Collection<U> | AccessorArrayLike<U>, predicate2: Predicate<U> ): number;
```
You can definitely simplify this defn with `InputArray`.

Also, you stopped with predicate pairs. I suggest going up to 5 array/predicate pairs (i.e., 10 arguments). Then, for your last overload you need to use a rest parameter which is a union of `InputArray | Predicate`, so that people can provide more than 10 args and not have the compiler complain.

kgryte commented on lib/node_modules/@stdlib/array/base/count-ifs/docs/types/index.d.ts:
```diff
@@ -0,0 +1,149 @@
+* var n = countIfs( x0, predicate0, x1, predicate1 );
+* // returns 2
+*/
+declare function countIfs<U = unknown>( x0: Collection<U> | AccessorArrayLike<U>, predicate0: Predicate<U>, x1: Collection<U> | AccessorArrayLike<U>, predicate1: Predicate<U> ): number;
```
This isn't correct. `x0` may have type `U`, but `x1` should be allowed to have type `V`. In short, each array should be able to have its own type.

kgryte commented on lib/node_modules/@stdlib/array/base/count-ifs/lib/main.js:
```diff
@@ -0,0 +1,122 @@
+	len = arrays[ 0 ].data.length;
+	for ( i = 0; i < len; i++ ) {
+		flg = true;
+		for ( j = 0; j < arrays.length; j++ ) {
```
This isn't a great implementation. (a) You're not cache-wise here, hopping around between arrays. (b) Calling `internal` is not great here, as you need to unwind properties each time. (c) You're not differentiating between accessor and indexed arrays.

Instead, I suggest

```javascript
len = arguments[ 0 ].length;
tmp = filled( true, len );
for ( i = 0; i < arrays.length; i++ ) {
	x = arrays[ i ];
	if ( x.accessorProtocol ) {
		accessors( x, predicates[ i ], tmp );
	} else {
		indexed( x.data, predicates[ i ], tmp );
	}
}
return countTruthy( tmp );
```

with

```javascript
function accessors( obj, predicate, out ) {
	var buf;
	var get;
	var i;

	buf = obj.data;
	get = obj.accessors[ 0 ];
	for ( i = 0; i < buf.length; i++ ) {
		if ( out [ i ] ) {
			out[ i ] = Boolean( predicate( get( buf, i ), i, buf ) );
		}
	}
	return out;
}

function indexed( x, predicate, out ) {
	var i;
	for ( i = 0; i < x.length; i++ ) {
		if ( out[ i ] ) {
			out[ i ] = Boolean( predicate( x[ i ], i, x ) );
		}
	}
	return out;
}
```

kgryte replied:
Your implementation does have the advantage of breaking early if an array's predicate returns falsy for an invocation, which is superior when working with 1 or 2 arrays. But in the first scenario, we already have `array/base/count-if`. For the general case, my suspicion is that it's more beneficial to consume each array in its entirety before moving on to the next.

kgryte replied:
Actually...maybe I have a better implementation. One second.

kgryte replied:
Okay. Here is the implementation which combines the ideas of both implementations...

```javascript
var zeroTo = require( '@stdlib/array/base/zero-to' );

// ...

len = arguments[ 0 ].length;
idx = zeroTo( len );
for ( i = 0; i < arrays.length; i++ ) {
	x = arrays[ i ];
	if ( x.accessorProtocol ) {
		idx = accessors( x, idx, predicates[ i ] );
	} else {
		idx = indexed( x.data, idx, predicates[ i ] );
	}
}
return idx.length;
```

with

```javascript
function accessors( obj, idx, predicate ) {
	var buf;
	var get;
	var i;
	var j;
	var k;

	buf = obj.data;
	get = obj.accessors[ 0 ];

	k = 0;
	for ( i = 0; i < idx.length; i++ ) {
		j = idx[ i ];
		if ( predicate( get( buf, j ), j, buf ) ) {
			idx[ k ] = j;
			k += 1;
		}
	}
	idx.length = k;
	return idx;
}

function indexed( x, idx, predicate ) {
	var i;
	var j;
	var k;

	k = 0;
	for ( i = 0; i < idx.length; i++ ) {
		j = idx[ i ];
		if ( predicate( x[ j ], j, x ) ) {
			idx[ k ] = j;
			k += 1;
		}
	}
	idx.length = k;
	return idx;
}
```

In short, we can keep track of a list of indices for which we have had all truthy return values thus far. For each array, we subsequently compact that list of indices and only invoke subsequent predicate functions for a reduced subset of indices.

headlessNode committed:
docs: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: incorrect signature

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: error due to default unknown type

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented on lib/node_modules/@stdlib/array/base/count-ifs/docs/types/index.d.ts:
```diff
@@ -0,0 +1,243 @@
+* var n = countIfs( x0, predicate0, x1, predicate1 );
+* // returns 2
+*/
+declare function countIfs<T, U extends InputArray<T>>( x0: U, predicate0: Predicate<T,U>, x1: U, predicate1: Predicate<T,U> ): number;
```
@kgryte I removed the `T = unknown` here and below because it was causing an error.

kgryte replied:
When you use `T = unknown`, you also need to set a default for `U`. I'll make the change.

kgryte replied:
Updated the declarations. Take a look for future reference.

kgryte committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: update descriptions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---
